Towing - Fix some issues

Dystopian opened the pull request:
**When merged this pull request will:**
- add multiple towing;
- add JIP compatibility;
- add towing to ships ~and helicopters~;
- add rope to boats inventory;
- add hook also to parent;
- add `Detach` actions to both towing/towed vehicle;
- replace hook model with invisible one;
- fix interaction with dead vehicle;
- fix immediate rope attachment when click `Attach tow rope` menu if player is close to vehicle;
- fix MP issue when rope looks detached on other clients;
- fix MP issue when you attach rope to non-local vehicle;
- fix MP issue when one player attaches vehicles and other player tows;
- fix MP issue when "tow" client is disconnected and rope is broken.

Multiple towing means either several ropes between parent and child or one parent with several children one by one and parallel.

`setTowParent` is not JIP compatible so I added workaround.

It turned out that ships can tow and can be towed, helicopters can be towed as well (at least those with wheels).

With parent hook rope can be detached from parent too.

Current hook model glows very much. This can be a problem at night when you want to hide vehicles. I didn't find more suitable model and used empty one which is OK. There is `\A3\Data_f\Hook\Hook_F.p3d` which is good but it could be weird for many vehicles.

`ropeAttachTo` and `ropeDetach` commands must be run where parent and rope are local.
`setTowParent` must be run on all clients.
Current implementation runs `Deleted` and `RopeBreak` EHs on client where rope was attached. When client disconnects EHs aren't triggered. The PR fixes it.

Dystopian commented:
Ready for review.

I didn't find a way to fix possible rope attachment inside vehicles resulting impossible to detach it. I added parent hook as partial workaround but if anybody knows a better way to fix please tell me.

Dystopian commented:
After #8963 and this PR rope can also be added to boats inventory. I can't really decide if it's needed.

dedmen commented on addons/towing/functions/fnc_attachRopePFH.sqf:
```diff
@@ -40,11 +40,30 @@ if (_intersections isNotEqualTo []) then {
 
     if (GVAR(canAttach)) then {
-        TRACE_4("can attach",_target,_intersectObject,_ignoreParent,_ignoreRope);
+        // TRACE_4("can attach",_target,_intersectObject,_ignoreParent,_ignoreRope);
```
Why remove debug code? its not present in release builds anyway

Dystopian replied:
When you enable debug it spams a really lot of lines to rpt. I disabled (not removed) only most annoying PFH debug, it can be easily enabled if needed.

dedmen commented on addons/towing/functions/fnc_isSuitableSimulation.sqf:
```diff
@@ -19,8 +19,7 @@ params ["_target"];
 // Biki lies, you can both tow and tow as either TankX or CarX
-(toLower _simulationType) in ["tankx", "carx"]
-
+(toLower _simulationType) in ["tankx", "carx", "shipx", "helicopterrtd"]
```
Ship and Helicopter have no support for setTowParent, the wiki is correct on that.

But ships don't have wheel brakes and steering (which is what setTowParent does)
And RTD helicopter also does some special funky stuff which is why it works, though does it also work with advaned helisim thing turned off?

Dystopian replied:
I only tested with disabled AFM. Can't test with enabled one for several weeks. So I hope someone will test it while reviewing.

Dystopian commented:
To fix possible rope attachment inside vehicles resulting impossible to detach it I added `Detach` actions on each towing/towed vehicle.

I tested towing helicopters with enabled AFM and it works OK as with disabled one. But I disabled towed helicopters because I don't think helicopters have suitable towing points (except landing gear maybe). Naturally this can be discussed.

I added rope to boats inventory because it looks quite appropriate.

rautamiekka commented on addons/towing/functions/fnc_attachVehicles.sqf:
```diff
@@ -0,0 +1,65 @@
+/*
+ * Author: Dystopian
+ * Attaches child to parent vehicle.
+ * Run globally.
```
Would be much better if the fnc did the 'Run globally' part automatically.

Dystopian replied:
Didn't get you. Do you mean I should replace one global event and one function with 3 events and its own functions (1 for parent owner, 1 global `setTowParent` run and 1 for server)?

rautamiekka replied:
I mean the script is changed to `Must be run on the server` and the Server runs it globally, otherwise `remoteExec`ed.

Dystopian replied:
Still can't understand how it could be better to move script executing to server and split 1 global event to 3 events (1 of those is global anyway) and functions. In **any** case we have to run globally `setTowParent` part. I don't see any profit from function splitting.

Dystopian commented:
> Whats the status on this?

This PR is quite large and should be tested well by ACE devs. So we have to wait with patience.

johnb432 commented on addons/towing/functions/fnc_startTow.sqf:
```diff
@@ -30,6 +30,6 @@ _unit removeItem _ropeClass;
 
 GVAR(cancel) = false;
 GVAR(canAttach) = false;
+GVAR(onMouseButtonDownEHID) = ["MouseButtonDown", LINKFUNC(onMouseButtonDown)] call CBA_fnc_addDisplayHandler;
```
Why add it and remove it every time? What was wrong with the old method?

Dystopian replied:
`- fix immediate rope attachment when click Attach tow rope menu if player is close to vehicle;`
see also the 1st commit in this PR 0394bcf1c12865a19108efe65f97575fdef56344 (actually reason of PR)

johnb432 commented on addons/towing/functions/fnc_towStateMachinePFH.sqf:
```diff
@@ -35,14 +35,16 @@ if (_exitCondition && {_state < TOW_STATE_CANCEL}) then {
+            // can't use unit hand because rope doesn't change position when hand moving
+            // can't use createVehicleLocal because rope can be non-local (like parent) and it must be attached to global vehicle
+            GVAR(helper) = createVehicle [QGVAR(helper), [0, 0, 0], [], 0, "CAN_COLLIDE"];
+            GVAR(helper) attachTo [_unit, [0,0,0], "LeftHand", true];
+            _rope = ropeCreate [_parent, _parent worldToModelVisual ASLtoAGL getPosASLVisual GVAR(attachHelper), GVAR(helper), [0,0,0], _length];
```
```suggestion
            GVAR(helper) attachTo [_unit, [0, 0, 0], "LeftHand", true];
            _rope = ropeCreate [_parent, _parent worldToModelVisual ASLtoAGL getPosASLVisual GVAR(attachHelper), GVAR(helper), [0, 0, 0], _length];
```

johnb432 commented on addons/towing/functions/fnc_detachRope.sqf:
```diff
@@ -0,0 +1,63 @@
+} forEach [_hook, _hookParent];
+
+// cleanup object variables and EHs only if function isn't called from Deleted EH
+if (isNull _deletedObject || {_parent isNotEqualTo _deletedObject}) then {
```
Just to make it consistent with the other instances:
```suggestion
if (isNull _deletedObject || {_parent != _deletedObject}) then {
```

Dystopian replied:
We deal here with object which has just been deleted. I can't test carefully your code next 2 weeks while I was testing my code since I created the PR. So sorry I can't accept this change until testing.

johnb432 commented on addons/towing/functions/fnc_detachRope.sqf:
```diff
@@ -0,0 +1,63 @@
+        [QGVAR(cleanupParent), _parent] call CBA_fnc_serverEvent;
+    };
+};
+if (isNull _deletedObject || {_child isNotEqualTo _deletedObject}) then {
```
```suggestion
if (isNull _deletedObject || {_child != _deletedObject}) then {
```

johnb432 reviewed (commented):
I haven't tested it yet, but I will soon.

rautamiekka commented on addons/towing/functions/fnc_detachRope.sqf:
```diff
@@ -0,0 +1,63 @@
+
+ropeDestroy _rope; // can run on client
+
+if (!isNull _unit) then {
+    TRACE_1("rope",_ropeClass);
+    if (!isNull _unit && {_ropeClass isNotEqualTo ""}) then {
+        [_unit, _ropeClass, true] call CBA_fnc_addItem;
+    };
+};
```
I doubt the lazy eval saves anything here either way.

johnb432 commented on addons/towing/XEH_postInit.sqf:
```diff
@@ -1,14 +1,39 @@
+    };
 }] call CBA_fnc_addEventHandler;
 
+addMissionEventHandler ["PlayerConnected", {
+    if (GVAR(allChildren) isEqualTo []) exitWith {};
+    params ["", "", "", "_jip", "_owner"];
+    if (!_jip) exitWith {};
+    TRACE_2("pushing children",_owner,GVAR(allChildren));
+    [QGVAR(setTowParentAllChildren), [GVAR(allChildren)], _owner] call CBA_fnc_ownerEvent;
+}];
```
I mean the setup as a whole. Why go with an EH `connected` structure instead of an approach that would use `CBA_fnc_globalEventJIP`? I'm just curious.

Dystopian replied:
If I understand you correctly, for `CBA_fnc_globalEventJIP` I would need make `GVAR(allChildren)` public which is now local to server. So I push it from server only when JIP client is connected.

johnb432 replied:
The implementation I had in mind was similar to the old one, where you had a `QGVAR(setTowParent)` EH. Instead of calling it on the target only though, I would have called it as global and JIP, like so:

`[QGVAR(setTowParent), [_child, _parent], _parent] call CBA_fnc_targetEvent;`

would become

`[QGVAR(setTowParent), [_child, _parent]] call CBA_fnc_globalEventJIP;`

The JIP event would be removed upon deletion of the object in question or when the object is no longer being towed.

matthawkins2 commented:
Also hoping to see this PR go through. Towing is completely inop in both singleplayer and multiplayer in 3.16

LinkIsGrim commented:
🍌
